src/core/abr/guess_based_chooser.ts

```diff
@@ -0,0 +1,303 @@
+  ) : boolean {
+    return isFinite(bufferGap) && bufferGap >= 2.5 &&
+           performance.now() > this._blockGuessesUntil &&
+           scoreData[1] === ScoreConfidenceLevel.HIGH &&
+           scoreData[0] / speed > 1.1;
+  }
+
+  /**
+   * Returns `true` if the pending guess of `lastGuess` seems to not
+   * be maintainable and as such should be stopped.
+   * @param {Object} lastGuess
+   * @param {Array} scoreData
+   * @param {number} bufferGap
+   * @param {Array.<Object>} requests
+   * @returns {boolean}
+   */
+  private _shouldStopGuess(
+    lastGuess : Representation,
+    scoreData : [number, ScoreConfidenceLevel] | undefined,
+    bufferGap : number,
+    requests : IRequestInfo[]
+  ) : boolean {
+    if (scoreData !== undefined && scoreData[0] < 1.05) {
+      return true;
+    } else if ((scoreData === undefined || scoreData[0] < 1.2) && bufferGap < 0.6) {
+      return true;
+    }
+
+    const guessedRepresentationRequests = requests.filter(req => {
+      return req.content.representation.id === lastGuess.id;
+    });
+
+    const now = performance.now();
+    for (let i = 0; i < guessedRepresentationRequests.length; i++) {
+      const req = guessedRepresentationRequests[i];
+      const requestElapsedTime = now - req.requestTimestamp;
+      if (req.content.segment.isInit) {
+        if (requestElapsedTime > 1000) {
+          return true;
+        }
+      } else if (requestElapsedTime > req.content.segment.duration * 1000 + 200) {
+        return true;
+      } else {
+        const fastBw = estimateRequestBandwidth(req);
+        if (fastBw !== undefined && fastBw < lastGuess.bitrate * 0.8) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private _isLastGuessValidated(
+    lastGuess : Representation,
+    incomingBestBitrate : number,
+    scoreData : [number, ScoreConfidenceLevel] | undefined
+  ) : boolean {
+    if (scoreData !== undefined &&
+        scoreData[1] === ScoreConfidenceLevel.HIGH &&
+        scoreData[0] > 1.5)
+    {
+      return true;
+    }
+    return incomingBestBitrate >= lastGuess.bitrate &&
+      (this._lastMaintanableBitrate === null ||
+        this._lastMaintanableBitrate < lastGuess.bitrate);
+  }
+}
+
+/**
+ * From the array of Representations given, returns the Representation with a
+ * bitrate immediately superior.
+ * Returns `null` if that "next" Representation is not found.
+ * @param {Array.<Object>} representations
+ * @param {Object} representation
+ * @returns {Object|null}
+ */
+function getNextRepresentation(
+  representations : Representation[],
+  currentRepresentation : Representation
+) : Representation | null {
+  const len = representations.length;
+  let index = arrayFindIndex(representations,
+                             ({ id }) => id === currentRepresentation.id);
+  if (index < 0) {
+    log.error("ABR: Current Representation not found.");
+    return null;
+  }
+
+  while (++index < len) {
```

Comment:
Just to be sure to understand the logic here, it is only valid when the array of representation is sorted by bitrate from the the lowest bitrate to the bigger bitrate?

Reply:
Oh yes you're right, this is a guarantee we enforce everywhere to avoid doing `sort` at every place but we may want to document this.

Reply:
Done, I added this rule at multiple place in the documentation.

Is that sufficient in your point of view?